snowfall/training/compute_expected_times.py

```diff
@@ -0,0 +1,231 @@
+    expected_sum_per_row = torch.ones_like(sum_per_row)
+    assert torch.allclose(sum_per_row, expected_sum_per_row)
+
+    frame_idx = torch.arange(paths_shape.num_elements(),
+                             device=device) - k2.index(path_starts,
+                                                       paths_shape.row_ids(1))
+
+    # TODO(fangjun): we can swap `rows` and `cols`
+    # while creating `pathframe_to_pathphone` so that
+    # `t()` can be omitted here.
+    weighted_occupation = torch.sparse.mm(
+        pathframe_to_pathphone.t(),
+        frame_idx.unsqueeze(-1).to(pathframe_to_pathphone.dtype))
+
+    # sum over columns to get the total occupation
+    # Use `to_dense()` here because PyTorch does not
+    # support div(dense_matrix, sparse_matrix)
+    total_occupation = torch.sparse.sum(pathframe_to_pathphone,
+                                        dim=0).to_dense()
+
+    # TODO(fangjun): remove print
+    print('total_occupation[:50]\n', total_occupation[:50])
+
+    eps = torch.finfo(total_occupation.dtype).eps
+    expected_times = weighted_occupation.squeeze() / (total_occupation + eps)
+
+    # Number of `pathphone_idx`'s should be even
+    assert expected_times.shape[0] & 1 == 0
+
+    # Replace the expected_times for `pathphone_idx`'s which are even with
+    # the average expected times of two neighboring phones
+    #
+    # Even `pathphone_idx`'s belong to epsilon self-loops.
+    expected_times[2::2] = (expected_times[1:-1:2] +
+                            expected_times[3::2]) * 0.5
+    expected_times[0] = 0
+
+    # TODO(fangjun): we can remove the columns of even pathphone_idx
+    # while constructing `pathframe_to_pathphone`, which can save about
+    # half computation time in `torch.sparse.mm`.
+
+    frame_idx_low = torch.floor(expected_times)
+    frame_idx_high = torch.ceil(expected_times)
+
+    low_scale = frame_idx_high - expected_times
+    high_scale = 1 - low_scale
+    offset = k2.index(seq_starts, pathphone_idx_to_seq)
+
+    low = frame_idx_low + offset
+    high = frame_idx_high + offset
+
+    low_scores = k2.index(dense_fsa_vec.scores, low.to(torch.int32))
+    high_scores = k2.index(dense_fsa_vec.scores, high.to(torch.int32))
+
+    embedding_scores = low_scores * low_scale.unsqueeze(
+        -1) + high_scores * high_scale.unsqueeze(-1)
+
+    # arc entering the final state have phone == -1.
+    # Increment it so that 0 represents EOS
+    pathphones += 1
+    num_classes = max_phone_id + 2  # +1 for the epsilon, +1 for EOS
+    print(pathphones.max(), pathphones.min(), num_classes)
+
+    # TODO(fangjun): do we need to build our own one_hot
+    # that supports dtype == torch.int32
+    embedding_phones = torch.nn.functional.one_hot(
+        pathphones.long(), num_classes=num_classes).to(device)
+
+    print(embedding_scores.shape)
+    print(embedding_phones.shape)
+    print(expected_times.unsqueeze(-1).shape)
+    embeddings = torch.cat(
```

Comment:
@danpovey 

Now we have `embeddings`. Can you explain how we use it?

Since the `embeddings` is a `torch.Tensor`, do we need to attach
a ragged shape to it?

Reply:
Great!!
Yes, they will have the same ragged shape as the pathphone_idx.
Basically we will be doing another pass of MMI training, except instead of summing the `tot_scores` we will compute a weighted sum where the weight is (num_copies_this_path / num_paths_this_seq).  For the numerator, the graphs are the same as we used in the 1st pass except we'll need the b_to_a_map or whatever it's called here.  For the denominator, the graphs are also the same.
Obviously the model is different though because it has a different input dim.  We'll have to create another model.  Can be a fairly small and simple one for now, e.g. with 1-d convolutions like our basic TDNN model.

Reply:
.. and this model, of course, won't do any subsampling.

Reply:
Thanks, will try it.

Reply:
... What we expect  is that, because the model sees the embeddings of the phones recognized in the 1st pass as input, it will typically give a fairly high weight to those 1st-pass phones, so that in most cases the 2nd-pass predictions will be "aligned" with those from the 1st pass.  However in cases where the 1st pass had deletion errors it will have to  make use of the 'epsilon' positions to insert phones between the 1st-pass decoding outputs.

In a few cases, when the 1st-pass output was too short by more than a factor of 2, we will get search errors.  I believe the current code should, appropriately, reject those utterances and not train on them (scores should be infinity).

Reply:
@danpovey 
I have two doubts about the implementation.

(1) The `embeddings` has a ragged shape `[paths][pathpone_idx]` but the network input
has to be a tensor with a regular shape. Different paths may have a different number of 
`pathphone_idx`'s. Should we zero pad the embeddings so that the `embeddings` for
each path has the same shape?  I think we can use the length information before padding
as `duration` while creating the `DenseFsaVec`.

(2)
> For the numerator, the graphs are the same as we used in the 1st pass except we'll need the b_to_a_map or whatever it's called here.

There is an `a_to_b_map` in `intersect_dense` but since each sequence may have multiple paths so I think `intersect_dense` is not applicable here.

There is a `b_to_a_map` in `intersect_device`, but it accepts two `Fsa` as inputs. To use it, we have to convert
`DenseFsaVec` to `Fsa`. 

Do you think the conversion is fine or do we need to extend `intersect_dense` to support `b_to_a_map`?

Thanks!

Reply:
> @danpovey
> I have two doubts about the implementation.
> 
> (1) The `embeddings` has a ragged shape `[paths][pathpone_idx]` but the network input
> has to be a tensor with a regular shape. Different paths may have a different number of
> `pathphone_idx`'s. Should we zero pad the embeddings so that the `embeddings` for
> each path has the same shape? I think we can use the length information before padding
> as `duration` while creating the `DenseFsaVec`.

Yes, that makes sense.  Eventually we can think about packing, but for now I don't think it's  necessary.


> > For the numerator, the graphs are the same as we used in the 1st pass except we'll need the b_to_a_map or whatever it's called here.
> 
> There is an `a_to_b_map` in `intersect_dense` but since each sequence may have multiple paths so I think `intersect_dense` is not applicable here.

Hm, that's a hassle... for now I think it will be easiest to just index the k2.Fsa using the b_to_a_map and use the current intersect_dense code.

Reply:
Yes, that is another alternative. Thanks.

Reply:
> For the numerator, the graphs are the same as we used in the 1st pass except we'll need the b_to_a_map or whatever it's called here. For the denominator, the graphs are also the same.

> Hm, that's a hassle... for now I think it will be easiest to just index the k2.Fsa using the b_to_a_map and use the current intersect_dense code.

It turns out it will use lots of memory if replicating `num_graph` and `den_graph`
for as many times as `num_paths`, which results in out-of-memory in `intersect_dense`.
I have reduced `max_frames` and the OOM remains.

Reply:
I am using `num_paths=10` while invoking `random_paths` as a workaround.